src/codeu/chat/util/PasswordStorage.java

```diff
@@ -0,0 +1,241 @@
+                ex
+            );
+        }
+
+        if (iterations < 1) {
+            throw new InvalidHashException(
+                "Invalid number of iterations. Must be >= 1."
+            );
+        }
+
+
+        byte[] salt = null;
+        try {
+            salt = fromBase64(params[SALT_INDEX]);
+        } catch (IllegalArgumentException ex) {
+            throw new InvalidHashException(
+                "Base64 decoding of salt failed.",
+                ex
+            );
+        }
+
+        byte[] hash = null;
+        try {
+            hash = fromBase64(params[PBKDF2_INDEX]);
+        } catch (IllegalArgumentException ex) {
+            throw new InvalidHashException(
+                "Base64 decoding of pbkdf2 output failed.",
+                ex
+            );
+        }
+
+
+        int storedHashSize = 0;
+        try {
+            storedHashSize = Integer.parseInt(params[HASH_SIZE_INDEX]);
+        } catch (NumberFormatException ex) {
+            throw new InvalidHashException(
+                "Could not parse the hash size as an integer.",
+                ex
+            );
+        }
+
+        if (storedHashSize != hash.length) {
+            throw new InvalidHashException(
+                "Hash length doesn't match stored hash length."
+            );
+        }
+
+        // Compute the hash of the provided password, using the same salt, 
+        // iteration count, and hash length
+        byte[] testHash = pbkdf2(password, salt, iterations, hash.length);
+        // Compare the hashes in constant time. The password is correct if
+        // both hashes match.
+        return slowEquals(hash, testHash);
+    }
+
+    private static boolean slowEquals(byte[] a, byte[] b)
+    {
+        int diff = a.length ^ b.length;
+        for(int i = 0; i < a.length && i < b.length; i++)
+            diff |= a[i] ^ b[i];
+        return diff == 0;
+    }
+
+    private static byte[] pbkdf2(char[] password, byte[] salt, int iterations, int bytes)
+        throws CannotPerformOperationException
+    {
+        try {
+            PBEKeySpec spec = new PBEKeySpec(password, salt, iterations, bytes * 8);
+            SecretKeyFactory skf = SecretKeyFactory.getInstance(PBKDF2_ALGORITHM);
+            return skf.generateSecret(spec).getEncoded();
+        } catch (NoSuchAlgorithmException ex) {
+            throw new CannotPerformOperationException(
+                "Hash algorithm not supported.", ex);
+        } catch (InvalidKeySpecException ex) {
+            throw new CannotPerformOperationException(
+                "Invalid key spec.", ex);
+        }
+    }
+
+    private static byte[] fromBase64(String hex)
+        throws IllegalArgumentException
+    {
+        return DatatypeConverter.parseBase64Binary(hex);
+    }
+
+    private static String toBase64(byte[] array)
+    {
+        return DatatypeConverter.printBase64Binary(array);
+    }
+
+    /* My added methods */
+
+    public static String setPassword() {
```

Comment:
see my comment below on authenticate(..) first.
Similarly here: better design is 1st ask the context if the user is a new Admin
```
if (context.isUserNewAdmin(userId)) {
  // ask user to enter new password.
  context.storePassword(userId, enteredPassword);
} else if (context.isUserAdmin(userId) {
  // ask user to enter their password.
  shouldContinue = context.authenticateUser(enteredPassword, userId);
  if (!shouldContinue) {
    // print "Incorrect password".
    // exit this function.
  }
}
// if we are here this means it's a normal user or authentication was successful.
createUserPanel(...)
}
```